File: composeApp/src/androidMain/kotlin/com/muralex/myapp/resources/StringRefResolver.kt
package com.muralex.myapp.resources

import androidx.compose.runtime.Composable
import com.muralex.myapp.R
import com.muralex.myapp.viewmodel.resources.SharedRes
import com.muralex.myapp.viewmodel.resources.StringRef
import com.muralex.myapp.viewmodel.resources.StringRefWithArgs

object StringRefResolver {
    @Composable
    fun resolve(ref: StringRef): String = when (ref) {
        SharedRes.Strings.settings_theme_title -> stringRes(R.string.settings_theme_title)
        SharedRes.Strings.settings_theme_option_system -> stringRes(R.string.settings_theme_option_system)
        SharedRes.Strings.settings_theme_option_dark -> stringRes(R.string.settings_theme_option_dark)
        SharedRes.Strings.settings_theme_option_light -> stringRes(R.string.settings_theme_option_light)
        SharedRes.Strings.settings_theme_dialog_title -> stringRes(R.string.settings_theme_dialog_title)
        SharedRes.Strings.settings_favorite_swipe_title -> stringRes(R.string.settings_favorite_swipe_title)
        SharedRes.Strings.settings_favorite_swipe_summaryOn -> stringRes(R.string.settings_favorite_swipe_summaryOn)
        SharedRes.Strings.settings_favorite_swipe_summaryOff -> stringRes(R.string.settings_favorite_swipe_summaryOff)
        SharedRes.Strings.settings_sync_title -> stringRes(R.string.settings_sync_title)
        SharedRes.Strings.settings_sync_summary -> stringRes(R.string.settings_sync_summary)
        SharedRes.Strings.settings_sync_dialog_title -> stringRes(R.string.settings_sync_dialog_title)
        SharedRes.Strings.settings_sync_dialog_message -> stringRes(R.string.settings_sync_dialog_message)
        SharedRes.Strings.settings_category_interface -> stringRes(R.string.settings_category_interface)
        SharedRes.Strings.settings_category_data -> stringRes(R.string.settings_category_data)
        SharedRes.Strings.settings_appversion_title -> stringRes(R.string.settings_appversion_title)
        SharedRes.Strings.settings_appversion_summary -> stringRes(R.string.settings_appversion_summary)
        SharedRes.Strings.settings_deviceinfo_title -> stringRes(R.string.settings_deviceinfo_title)
        SharedRes.Strings.settings_deviceinfo_summary -> stringRes(R.string.settings_deviceinfo_summary)
        SharedRes.Strings.settings_sync_in_progress -> stringRes(R.string.settings_sync_in_progress)
        else -> ref.simpleName()
    }

    @Composable
    fun resolve(ref: StringRefWithArgs, arg: String): String = when (ref.ref) {
        SharedRes.Strings.settings_sync_last_formatted ->
            stringRes(R.string.settings_sync_last, arg)

        SharedRes.Strings.settings_sync_success_formatted ->
            stringRes(R.string.settings_sync_success, arg)

        SharedRes.Strings.settings_sync_failed_formatted ->
            stringRes(R.string.settings_sync_failed, arg)

        else -> ref.ref.simpleName()
    }

    @Composable
    fun resolve(ref: StringRefWithArgs, vararg args: String): String = when (ref.ref) {
        SharedRes.Strings.settings_summary_current_formatted ->
            stringRes(R.string.settings_summary_current, *args)

        else -> ref.ref.simpleName()
    }
}

File: iosApp/iosApp/views/resources/StringRefResolver.swift
 

import SwiftUI
import Shared


extension SharedRes.Strings {
    func asString() -> String {
        switch self {
        case .settingsThemeTitle: return "Theme"
        case .settingsThemeOptionSystem: return "System"
        case .settingsThemeOptionDark: return "Dark"
        case .settingsThemeOptionLight: return "Light"
        case .settingsThemeDialogTitle: return "Choose Theme"
        case .settingsFavoriteSwipeTitle: return "Swipe to remove"
        case .settingsFavoriteSwipeSummaryOn: return "Swipe enabled"
        case .settingsFavoriteSwipeSummaryOff: return "Swipe disabled"
        case .settingsSyncTitle: return "Content updates"
        case .settingsSyncSummary: return "Keep app content up to date"
        case .settingsSyncInProgress: return "Updating content…"
        case .settingsCategoryInterface: return "Interface"
        case .settingsCategoryData: return "Data"
        case .settingsAppversionTitle: return "App Version"
        case .settingsAppversionSummary: return "Version info"
        case .settingsDeviceinfoTitle: return "Device Info"
        case .settingsSyncDialogTitle: return "Check for updates?"
        case .settingsSyncDialogMessage: return "\nWe’ll refresh the app content with the latest available data.\n"
        default: return String(describing: self)
        }
    }
}

extension StringRef {
    func asString() -> String {
        if let ref = self as? SharedRes.__Strings {
            return ref.toSwiftEnum().asString()
        }
        return String(describing: self)
    }
}


extension StringRefWithArgs {
    func asStringWithArg(_ arg: String) -> String {
        switch self.ref {
        case SharedRes.Strings.settingsSummaryCurrentFmt: return "Current: \(arg)"
        case SharedRes.Strings.settingsSyncLastFmt: return "Last updated: \(arg)"
        case SharedRes.Strings.settingsSyncSuccessFmt: return "Content updated successfully\nLast updated: \(arg)"
        case SharedRes.Strings.settingsSyncFailedFmt: return "Update failed. Please try again.\nLast updated: \(arg)"
        default: return self.ref.simpleName()
        }
    }
}

extension FormattedText {
    func asString() -> String {
        switch self {
        case let t as FormattedText.Ref: return t.ref.asStringWithArg("")
        case let t as FormattedText.SimpleText: return t.textRef.asString()
        case let t as FormattedText.WithString: return t.ref.asStringWithArg(t.arg)
        case let t as FormattedText.WithRef: return t.ref.asStringWithArg(t.arg.asString())
        default: return self.ref.ref.simpleName()
        }
    }
    
    func asString(arg: String) -> String {
        switch self {
        case let t as FormattedText.Ref: return t.ref.asStringWithArg(arg)
        default: return self.ref.ref.simpleName()
        }
    }
}
